clients/filesystem-fuse/src/default_raw_filesystem.rs

```diff
@@ -0,0 +1,398 @@
+        {
+            let mut file_manager = self.file_entry_manager.write().await;
+            file_manager.remove(&join_file_path(&parent_file_entry.path, name));
+        }
+        Ok(())
+    }
+
+    async fn close_file(&self, _file_id: u64, fh: u64) -> crate::filesystem::Result<()> {
+        let opened_file = self
+            .opened_file_manager
+            .remove(fh)
+            .ok_or(Errno::from(libc::EBADF))?;
+        let mut file = opened_file.lock().await;
+        file.close().await
+    }
+
+    async fn read(
+        &self,
+        _file_id: u64,
+        fh: u64,
+        offset: u64,
+        size: u32,
+    ) -> crate::filesystem::Result<Bytes> {
+        let file_stat: FileStat;
+        let data = {
+            let opened_file = self
+                .opened_file_manager
+                .get(fh)
+                .ok_or(Errno::from(libc::EBADF))?;
+            let mut opened_file = opened_file.lock().await;
+            file_stat = opened_file.file_stat.clone();
+            opened_file.read(offset, size).await
+        };
+
+        // update the file atime
+        self.fs.set_attr(&file_stat.path, &file_stat, false).await?;
+
+        data
+    }
+
+    async fn write(
+        &self,
+        _file_id: u64,
+        fh: u64,
+        offset: u64,
+        data: &[u8],
+    ) -> crate::filesystem::Result<u32> {
+        let (len, file_stat) = {
+            let opened_file = self
+                .opened_file_manager
+                .get(fh)
+                .ok_or(Errno::from(libc::EBADF))?;
+            let mut opened_file = opened_file.lock().await;
+            let len = opened_file.write(offset, data).await;
+            (len, opened_file.file_stat.clone())
+        };
+
+        // update the file size, mtime and atime
+        self.fs.set_attr(&file_stat.path, &file_stat, false).await?;
+
+        len
+    }
+}
+
+/// File entry is represent the abstract file.
+#[derive(Debug, Clone)]
+struct FileEntry {
+    file_id: u64,
+    parent_file_id: u64,
+    path: String,
+}
+
+/// FileEntryManager is manage all the file entries in memory. it is used manger the file relationship and name mapping.
+struct FileEntryManager {
+    // file_id_map is a map of file_id to file entry.
+    file_id_map: HashMap<u64, FileEntry>,
+
+    // file_path_map is a map of file path to file entry.
+    file_path_map: HashMap<String, FileEntry>,
+}
+
+impl FileEntryManager {
+    fn new() -> Self {
+        Self {
+            file_id_map: HashMap::new(),
+            file_path_map: HashMap::new(),
+        }
+    }
+
+    fn get_file_entry_by_id(&self, file_id: u64) -> Option<FileEntry> {
+        self.file_id_map.get(&file_id).cloned()
+    }
+
+    fn get_file_entry_by_path(&self, path: &str) -> Option<FileEntry> {
+        self.file_path_map.get(path).cloned()
+    }
+
+    fn insert(&mut self, parent_file_id: u64, file_id: u64, path: &str) {
+        let file_entry = FileEntry {
```

Comment:
an improve way be use Box::new to avoid clone `FileEntry` object

Reply:
Boxed objects cannot be shared, that's like unique point, can't not use by multiple threads

Reply:
use Arc::new?

Reply:
That's not good